File: lab/c/matematica_e_operadores/contagem_de_um_a_n.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

void contagem1(int inicio, int fim, int passo) {
    if (passo <= 0) { printf("Passo invalido\nFIM!\n"); return; }
    for (int i = inicio; i <= fim; i += passo) {
        printf("%d ", i);
    }
    printf("FIM!\n");
}

void contagem2(int inicio, int fim, int passo) {
    if (passo <= 0) { printf("Passo invalido\nFIM!\n"); return; }
    int i;
    for (i = inicio; i <= fim; i += passo) {
        printf("%d ", i);
    }
    printf("FIM!\n");
}

void contagem_for(int N) {
    for (int i = 1; i <= N; i++) {
        printf("%d ", i);
    }
}

void contagem_while(int N) {
    int i = 1;
    while (i <= N) {
        printf("%d ", i);
        i++;
    }
}

void contagem_do_while(int N) {
    int i = 1;
    if (N < 1) { return; }        // evita imprimir nada se N < 1
    do {
        printf("%d ", i);
        i++;
    } while (i <= N);
}

// helper: lowercase in-place
void tolower_str(char *s) {
    for (; *s; s++) *s = (char)tolower((unsigned char)*s);
}

// Versão com if/else (equivalente ao seu contagem_simples)
void contagem_simples(int N, const char *loop) {
    char opt[32];
    snprintf(opt, sizeof(opt), "%s", loop);
    tolower_str(opt);

    if (strcmp(opt, "for") == 0) {
        contagem_for(N);
    } else if (strcmp(opt, "while") == 0) {
        contagem_while(N);
    } else if (strcmp(opt, "do-while") == 0 || strcmp(opt, "dowhile") == 0) {
        contagem_do_while(N);
    } else {
        printf("Loop invalido! Use 'for', 'while' ou 'do-while'.\n");
        printf("FIM!\n");
        return;
    }
    printf("FIM!\n");
}

// Versão “tabela” com ponteiros de função (equivalente ao seu contagem_simples2)
typedef void (*cont_fn)(int);

typedef struct {
    const char *nome;
    cont_fn fn;
} Entry;

void contagem_simples2(int N, const char *loop) {
    Entry table[] = {
        {"for",      contagem_for},
        {"while",    contagem_while},
        {"do-while", contagem_do_while},
        {"dowhile",  contagem_do_while},
    };
    char opt[32];
    snprintf(opt, sizeof(opt), "%s", loop);
    tolower_str(opt);

    int found = 0;
    for (int i = 0; i < (int)(sizeof(table)/sizeof(table[0])); i++) {
        if (strcmp(opt, table[i].nome) == 0) {
            table[i].fn(N);
            found = 1;
            break;
        }
    }
    if (!found) {
        printf("Loop invalido! Use 'for', 'while' ou 'do-while'.\n");
    }
    printf("FIM!\n");
}

int main(void) {
    // equivalentes de contagem1/2
    contagem1(1, 10, 1);
    contagem2(1, 10, 1);
    printf("\n");

    // equivalentes de contagem_simples
    contagem_simples(5, "for");
    contagem_simples(5, "while");
    contagem_simples(5, "dowhile");
    printf("\n");

    // equivalentes de contagem_simples2 (tabela de funções)
    contagem_simples2(5, "for");
    contagem_simples2(5, "while");
    contagem_simples2(5, "do-while");

    return 0;
}

File: lab/javascript/matematica_e_operadores/contagem_de_um_a_n.js
function contagem1(inicio, fim, passo) {
    let lista = [];
    while (inicio <= fim) {
        lista.push(inicio);
        inicio += passo;
    }
    lista.push("FIM!");
    console.log(lista.join(" "));
}


function contagem2(inicio, fim, passo) {
    let lista = [];
    for (let i = inicio; i <= fim; i += passo) {
        lista.push(i);
    }
    lista.push("FIM!");
    console.log(lista.join(" "));
}


function contagem_for(N){
    let lista = [];
    for (let i = 1; i <= N; i++) {
        lista.push(i);
    }
    return lista;
}


function contagem_while(N){
    let lista = [];
    let i = 1;
    while (i <= N) {
        lista.push(i);
        i++;
    }
    return lista;
}


function contagem_do_while(N){
    let lista = [];
    let i = 1;
    do {
        lista.push(i);
        i++;
    } while (i <= N);
    return lista;
}

function contagem_simples(N, loop='for'){
    loop = loop.toLowerCase();
    switch(loop){
        case 'for':
            retorno = contagem_for(N);
            break;
        case 'while':
            retorno = contagem_while(N);
            break;
        case 'do-while':
        case 'dowhile':
            retorno = contagem_do_while(N);
            break;
        default:
            retorno = "Loop inválido! Use 'for', 'while' ou 'do-while'.";
            console.log(retorno);
            console.log("FIM!");
            return;
    }
    retorno.push("FIM!");
    console.log(retorno.join(" "));
}

function contagem_simples2(N, loop='for'){
    let functions = {
        'for': contagem_for,
        'while': contagem_while,
        'do-while': contagem_do_while,
        'dowhile': contagem_do_while
    }

    if (functions[loop]){
        let retorno = functions[loop](N);
        retorno.push("FIM!");
        console.log(retorno.join(" "));
    } else {
        console.log("Loop inválido! Use 'for', 'while' ou 'do-while'.");
        console.log("FIM!");
    }
}

contagem1(1, 10, 1);
contagem2(1, 10, 1);

contagem_simples(5, "for")
contagem_simples(5, "while")
contagem_simples(5, "dowhile")

contagem_simples2(5, "for")
contagem_simples2(5, "while")
contagem_simples2(5, "do-while")
